yjit/src/codegen.rs

```diff
@@ -6737,8 +6735,9 @@ fn gen_opt_getconstant_path(
             RUBY_OFFSET_ICE_VALUE
         ));
 
-        // Push ic->entry->value
-        let stack_top = ctx.stack_push(Type::Unknown);
+        // Push ic->entry->value. We can assume a stable type for constants until a possible reassignment, which
+        // invalidates the IC entry and forces a recompile anyway.
+        let stack_top = ctx.stack_push(Type::from(unsafe { (*ice).value }));
```

Comment:
This code is pretty opaque, but IIUC, the ice entry is not initialized in this branch? @jhawthorn @XrXr

Reply:
Yeah, this is not safe to do. If we go down this code path, we don't invalidate on constant reassignment at all. It's also why we generate code to load from the cache instead of baking the VALUE into the generated like in the else branch. Maybe this is the cause of CI failures.

Reply:
Thanks for the review (and quick too!). I thought `rb_yjit_constant_ic_update` was doing some magic code invalidation voodoo but evidently it isn't doing what I thought it was doing. Since constants change type infrequently I guess there could be a way to do assume unchanging type, but I'll leave that for the people that know what they're doing 😄

Reply:
Hey, thanks for the patch. You wrote out your reasoning which made it easier to review. `rb_yjit_constant_ic_update` does do invalidation, but only for code generated in the `else` branch of `!unsafe { (*ice).ic_cref }.is_null()`. In that branch, we do lift the type of the cached constant value via `jit_putobject`. For readability maybe we ought to reorder the `if` and have the more common, non lexical scope case first 😄